pkg/metadata/sandbox.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metadata
+
+import (
+	"encoding/json"
+
+	"github.com/kubernetes-incubator/cri-containerd/pkg/metadata/store"
+
+	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+)
+
+// TODO(random-liu): Uncomment following code and handle versioning around
+// all marshal/unmarshal.
+// version and versionedSandboxMetadata is not used now, but should be used
+// in the future:
+// 1) Only versioned metadata should be written into the store;
+// 2) Only unversioned metadata should be returned to the user;
+// 3) A conversion function is needed to convert any supported versioned
+// metadata into unversioned metadata.
+
+// version is current version of sandbox metadata.
+const sandboxMetadataVersion = "v1" // nolint
+
+// versionedSandboxMetadata is the internal versioned sandbox metadata.
+// nolint
+type versionedSandboxMetadata struct {
+	// Version indicates the version of the versioned sandbox metadata.
+	Version string
+	SandboxMetadata
+}
+
+// SandboxMetadata is the unversioned sandbox metadata.
+type SandboxMetadata struct {
+	// ID is the sandbox id.
+	ID string
+	// Name is the sandbox name.
+	Name string
+	// Config is the CRI sandbox config.
+	Config *runtime.PodSandboxConfig
+	// CreatedAt is the created timestamp.
+	CreatedAt int64
+	// NetNS is the network namespace used by the sandbox.
+	NetNS string
+}
+
+// SandboxUpdateFunc is the function used to update SandboxMetadata.
+type SandboxUpdateFunc func(SandboxMetadata) (SandboxMetadata, error)
+
+// sandboxToStoreUpdateFunc generates a metadata store UpdateFunc from SandboxUpdateFunc.
+func sandboxToStoreUpdateFunc(u SandboxUpdateFunc) store.UpdateFunc {
+	return func(data []byte) ([]byte, error) {
+		meta := &SandboxMetadata{}
+		if err := json.Unmarshal(data, meta); err != nil {
+			return nil, err
+		}
+		newMeta, err := u(*meta)
+		if err != nil {
+			return nil, err
+		}
+		return json.Marshal(newMeta)
+	}
+}
+
+// SandboxStore is the store for metadata of all sandboxes.
+type SandboxStore interface {
```

Comment:
I don't think we'll have other implementations at this level. But it seems more clear to me to have an interface to tell the caller what operations are supported.